libsolidity/ast/TypeProvider.cpp

```diff
@@ -0,0 +1,395 @@
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libsolidity/ast/TypeProvider.h>
+#include <libdevcore/ArrayUtil.h>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/split.hpp>
+
+using namespace std;
+using namespace dev;
+using namespace solidity;
+
+template <size_t... N>
+constexpr std::array<IntegerType, sizeof...(N)> createIntegerTypes(IntegerType::Modifier _modifier, std::index_sequence<N...>)
+{
+	return makeArray<IntegerType>(IntegerType((static_cast<unsigned>(N) + 1) * 8, _modifier)...);
+}
+
+template <size_t... N>
+constexpr std::array<FixedBytesType, sizeof...(N)> createFixedBytesTypes(std::index_sequence<N...>)
+{
+	return makeArray<FixedBytesType>(FixedBytesType(static_cast<unsigned>(N) + 1)...);
+}
+
+BoolType const TypeProvider::m_boolType{};
+InaccessibleDynamicType const TypeProvider::m_inaccessibleDynamicType{};
+ArrayType const TypeProvider::m_bytesType{DataLocation::Storage, false};
+ArrayType const TypeProvider::m_bytesMemoryType{DataLocation::Memory, false};
+ArrayType const TypeProvider::m_stringType{DataLocation::Storage, true};
+ArrayType const TypeProvider::m_stringMemoryType{DataLocation::Memory, true};
+TupleType const TypeProvider::m_emptyTupleType{};
+AddressType const TypeProvider::m_payableAddressType{StateMutability::Payable};
+AddressType const TypeProvider::m_addressType{StateMutability::NonPayable};
+std::array<IntegerType, 32> const TypeProvider::m_intM{createIntegerTypes(IntegerType::Modifier::Signed, std::make_index_sequence<32>{})};
+std::array<IntegerType, 32> const TypeProvider::m_uintM{createIntegerTypes(IntegerType::Modifier::Unsigned, std::make_index_sequence<32>{})};
+std::array<FixedBytesType, 32> const TypeProvider::m_bytesM{createFixedBytesTypes(std::make_index_sequence<32>{})};
+std::array<MagicType, 4> const TypeProvider::m_magicTypes{
+	MagicType{MagicType::Kind::Block},
+	MagicType{MagicType::Kind::Message},
+	MagicType{MagicType::Kind::Transaction},
+	MagicType{MagicType::Kind::ABI}
+	// MetaType is stored separately
+};
+
+TypeProvider::TypeProvider()
+{
+}
+
+inline void clearCache(Type const& type)
+{
+	type.clearCache();
+}
+
+template <typename... Args>
+inline void clearCache(Type const& type, Args&... moreTypes)
+{
+	clearCache(type);
+	clearCache(moreTypes...);
+}
+
+template <typename Container>
+inline void clearAllCaches(Container& container)
+{
+	for_each(begin(container), end(container), [](Type const& t) { t.clearCache(); });
+}
+
+template <typename Container, typename... Args>
+inline void clearAllCaches(Container& types, Args&... more)
+{
+	clearAllCaches(types);
+	clearAllCaches(more...);
+}
+
+void TypeProvider::reset()
+{
+	clearCache(
```

Comment:
Whenever we add a new type, we have to also add it here. Wouldn't it be better to just re-create the TypeProvider?

Reply:
Recreating the type provider actually would implicitly clear that concern, yes. And that was my initial idea. However, I'd favor doing that in an isolated PR as a follow-up stage.